### PR TITLE
Json error format

### PR DESCRIPTION
**Pull request by Vbif**

### CODE REVIEW

File: VisualRust.Test/Build/RustcMessageJsonParserTest.cs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            const string output = "";
+            var messages = RustcMessageJsonParser.Parse(output).ToList();
+
+            Assert.AreEqual(0, messages.Count);
+        }
+
+        [Test]
+        public void FileNotFound()
+        {
+            const string output = "{\"message\":\"couldn't read \\\"main2.rs\\\": Не удается" +
```
Comment (Boddlnagg) [inline VisualRust.Test/Build/RustcMessageJsonParserTest.cs]: Does this mean that the test will fail when the system language is not Russian? That would be bad. Can you try to make this locale independent?

File: VisualRust.Test/VisualRust.Test.csproj [hunk 2]
```diff
@@ -117,7 +124,9 @@
       <Name>VisualRust.Project</Name>
     </ProjectReference>
   </ItemGroup>
-  <ItemGroup />
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
```
Comment (Boddlnagg) [inline VisualRust.Test/VisualRust.Test.csproj]: Do you have any idea where this is coming from / what it means?

File: VisualRust.Build/Message/Json/RustcMessageJsonParser.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace VisualRust.Build.Message.Json
+{
+    public static class RustcMessageJsonParser
+    {
+        private static readonly JsonSerializer serializer = new JsonSerializer();
+
+        public static IEnumerable<RustcMessageJson> Parse(String output)
+        {
+            var reader = new JsonTextReader(new StringReader(output)) {SupportMultipleContent = true};
```
Comment (Boddlnagg) [inline VisualRust.Build/Message/Json/RustcMessageJsonParser.cs]: I assume that the JSON deserialization keeps working if the format is changed (hopefully only backwards compatibly, i.e. by adding new fields), right?

File: VisualRust.Build/Rustc.cs [hunk 4]
```diff
@@ -182,8 +192,53 @@ public override bool Execute()
...
+        {
+            var process = CreateProcess(" --version");
+            process.Start();
+            process.WaitForExit();
+            return Version.ParseRustcOutput(process.StandardOutput.ReadToEnd());
+        }
+
         private bool ExecuteInner()
         {
+            var useJsonErrorFormat = rustcVersion.VersionMinor >= 11;
```
Comment (Boddlnagg) [inline VisualRust.Build/Rustc.cs]: Can you also check `rustcVersion.VersionMajor >= 1` ?  I don't think that anyone uses v0.11 anymore, but the check just seems ... incomplete.

**Comment by Boddlnagg** [comment]
Thanks! I added some comments inline. Apart from that, this looks great :)

@vosen: Are you still maintaining this project? Do you also want to have a look? And why is AppVeyor not set up to test PRs also?

File: VisualRust.Test/VisualRust.Test.csproj (hunk 2 above)
Comment (Vbif) [inline VisualRust.Test/VisualRust.Test.csproj]: Visual Studio continuesly add this line. There some disscusstions about this line https://github.com/nunit/nunit-vs-adapter/issues/51 http://stackoverflow.com/questions/18614342/what-is-service-include-in-a-csproj-file-for. We can remain this line or change project type.

File: VisualRust.Build/Message/Json/RustcMessageJsonParser.cs (hunk 3 above)
Comment (Vbif) [inline VisualRust.Build/Message/Json/RustcMessageJsonParser.cs]: I dont markup any of JSON fields as "required", so deserializer probably "works" on any input. All new fileds will be skipped.

File: VisualRust.Test/Build/RustcMessageJsonParserTest.cs (hunk 1 above)
Comment (Vbif) [inline VisualRust.Test/Build/RustcMessageJsonParserTest.cs]: These test just check basic mapping between JSON and RustcMessageJson class. So it is not depend on language, but I definitely need translate error message

**Comment by Boddlnagg** [comment]
I just tested this and found a few more problems:
- The spans are incorrect: When you call `LogError`/`LogWarning`, you should use `column_start`/`column_end` instead of `byte_start`/`byte_end` for the columns.
- The error `aborting due to previous error` should not be forwarded
- I think you have to add `Newtonsoft.Json.dll` as a component to the installer next to `VisualRust.Build.dll`. This is in `VisualRust.Setup\VisualRust.wxs`.

Besides that it seems to work :+1:

**Comment by Vbif** [comment]
I have a problem with building setup package. 

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		The "ResolveWixReferences" task could not be loaded from the assembly \WixTasks.dll. Could not load file or assembly 'file:///C:\WixTasks.dll' or one of its dependencies. Не удается найти указанный файл. Confirm that the <UsingTask> declaration is correct, that the assembly and all its dependencies are available, and that the task contains a public class that implements Microsoft.Build.Framework.ITask.	VisualRust.Setup	C:\Program Files (x86)\MSBuild\Microsoft\WiX\v3.x\wix2010.targets	735	

```

**Comment by Boddlnagg** [comment]
@Vbif What version of [WiX](http://wixtoolset.org/releases/) do you have installed? It works here with WiX 3.10.3

**Comment by Vbif** [comment]
v3.10.3.2904

**Comment by Boddlnagg** [comment]
Mine is v3.10.3.3007 .... and it looks like [this issue](https://github.com/wixtoolset/issues/issues/5291) was actually fixed there.

**Comment by Boddlnagg** [comment]
Will you be able to finish this? Otherwise I'm going to do it.
One more thing: You're checking for a version ≥ 1.11, but `--error-format=json` is only stabilized in 1.12, so you should change the version check to check for ≥ 1.12.

**Comment by Boddlnagg** [comment]
Thanks!

**Comment by vosen** [comment]
@Boddlnagg
yes, still interested in maintaining VR, just didn't have any free time lately. Conversely, are you? If so ass youself to the README.
I disable AppVeyor on PR, because I've jave been commiting faster than AppVeyor could test my changes (It could take an hour to have a run) and with NCrunch my master-breaking changes were mostly isolated to WiX (which AppVeyor didn't catch anyway) so I settled on tests after merge.

**Comment by Boddlnagg** [comment]
@vosen: I don't really have enough time to take over maintenance, either.
As far as CI tests for PRs are concerned: That sounds reasonable. It still would be useful to have them *sometimes* (like, on demand), but that's probably not possible?

**Comment by vosen** [comment]
Ill if look into it

**Pull request merged**